File: packages/core/src/evidence.ts
import type { EvidenceGrade, EvidenceUse } from "./types";

const rank: Record<EvidenceGrade, number> = {
  OFFICIAL_OBSERVED: 5,
  ENTITY_REPORTED: 4,
  THIRD_PARTY_VERIFIED: 3,
  TEXT_EXTRACTED: 2,
  MODEL_INFERRED: 1,
};

export function evidenceRank(grade: EvidenceGrade): number {
  return rank[grade];
}

export function allowedEvidenceUses(grade: EvidenceGrade): EvidenceUse[] {
  switch (grade) {
    case "OFFICIAL_OBSERVED":
    case "ENTITY_REPORTED":
    case "THIRD_PARTY_VERIFIED":
      return ["primary", "bounded", "retrieval"];
    case "TEXT_EXTRACTED":
    case "MODEL_INFERRED":
      return ["bounded", "retrieval"];
  }
}

export function canUseEvidence(
  grade: EvidenceGrade,
  use: EvidenceUse,
): boolean {
  return allowedEvidenceUses(grade).includes(use);
}

export function includedInBound(
  grade: EvidenceGrade,
  bound: "lower" | "central" | "upper",
): boolean {
  if (bound === "lower") {
    return grade === "OFFICIAL_OBSERVED" || grade === "ENTITY_REPORTED";
  }
  if (bound === "central") {
    return grade !== "TEXT_EXTRACTED" && grade !== "MODEL_INFERRED";
  }
  return true;
}

File: src/cascadelens/evidence.py
"""Evidence-grade policy used by all bounds."""

from __future__ import annotations

from typing import Literal

EvidenceGrade = Literal[
    "OFFICIAL_OBSERVED",
    "ENTITY_REPORTED",
    "THIRD_PARTY_VERIFIED",
    "TEXT_EXTRACTED",
    "MODEL_INFERRED",
]
Bound = Literal["lower", "central", "upper"]

EVIDENCE_GRADES: tuple[EvidenceGrade, ...] = (
    "OFFICIAL_OBSERVED",
    "ENTITY_REPORTED",
    "THIRD_PARTY_VERIFIED",
    "TEXT_EXTRACTED",
    "MODEL_INFERRED",
)

_RANK: dict[EvidenceGrade, int] = {
    "OFFICIAL_OBSERVED": 5,
    "ENTITY_REPORTED": 4,
    "THIRD_PARTY_VERIFIED": 3,
    "TEXT_EXTRACTED": 2,
    "MODEL_INFERRED": 1,
}


def evidence_rank(grade: EvidenceGrade) -> int:
    return _RANK[grade]


def allowed_uses(grade: EvidenceGrade) -> tuple[str, ...]:
    if grade in {"OFFICIAL_OBSERVED", "ENTITY_REPORTED", "THIRD_PARTY_VERIFIED"}:
        return ("primary", "bounded", "retrieval")
    return ("bounded", "retrieval")


def can_use(grade: EvidenceGrade, use: str) -> bool:
    return use in allowed_uses(grade)


def included_in_bound(grade: EvidenceGrade, bound: Bound) -> bool:
    if bound == "lower":
        return grade in {"OFFICIAL_OBSERVED", "ENTITY_REPORTED"}
    if bound == "central":
        return grade not in {"TEXT_EXTRACTED", "MODEL_INFERRED"}
    return True
